### DISCUSSION TITLE
How to use Qr Codes instead of barcodes

### DISCUSSION THREAD

**Discussion by mbappai** [alexandercerutti/passkit-generator#129] [Q&A] [answered]
My main question is to know how to configure and display a qr code instead of a barcode, and secondly, this might be a noob question but how am I going to download a stream from the frontend?

**Comment by alexandercerutti** [comment, marked as answer]
Hey there @mbappai,
for what concerns the QRCode, you can use [`setBarcodes`](https://github.com/alexandercerutti/passkit-generator/wiki/API-Documentation-Reference#setting-barcodes). If you look in the examples below in the paragraph I've linked you, you can see `PKBarcodeFormatQR`. That's the key to show a qrcode.

About downloading a stream from the frontend, assuming you are on a web page and not on an iOS application (both can be considered "frontends"), I don't actually know if there's a supported way on web to receive streams.

Instead I suggest you to use `.getAsBuffer()` and return the whole pass.

**Reply by mbappai** [reply]
Ahh I see! That's solves it for the qr code then I will check that out.

And as for the suggestion of using `.getAsBuffer()`, what type of format will the file come in? Because the intended behaviour on the frontend (web) will be to display a button that reads "Add to Apple Pass" which will then download the file () to their device (if it's supported).

**Reply by alexandercerutti** [reply]
Every pkpass is a zip file with a specific mime-type. You can find the possibilities here: https://github.com/alexandercerutti/passkit-generator/wiki/API-Documentation-Reference#mimetype-getter-only---inherited-from-bundle

I think that you can serve the file by using `.mimeType` and the device should automatically understand how to behave.

Understanding if the device is supported is a check you should do before rendering the button.

**Reply by mbappai** [reply]
I understand. I will try that and let you know how that goes. And yes I will have to check user device on my end and only show button when device supports a pass. Thanks a lot alexander for the quick response

**Reply by alexandercerutti** [reply]
You are welcome! I'll be waiting for your feedback 👍

**Reply by mbappai** [reply]
Hey Alex! So I was able to download the file after multiple trials (which i'll document in the blogpost ), but then Safari keeps telling me that it can't download the file. Below you'll find attached screenshots from my mobile device.

<img alt="" src="https://user-images.githubusercontent.com/26521423/225642928-f632eb0f-0690-4ab8-914d-bdbf387be2d0.PNG" width="300" />

<img alt="" src="https://user-images.githubusercontent.com/26521423/225642950-14d4b989-f60b-4c29-96f0-f7be460cca4b.PNG" width="300" />

So my question here is, is this issue caused by the downloaded file itself or maybe because of my phone version (ios 15.3)
PS: I saved the file with the extension of `.pkpass`

**Reply by alexandercerutti** [reply]
Uh, that's new to me. Maybe the second screenshot is about some kind of permission that hasn't been given, or maybe because it is the first time you are attempting to show a pass.

The first screenshot I'm not sure. You should attempt to debug through `Console.app` to understand if PKPass, perhaps returns any meaningful message. You can checkout the guide here: https://github.com/alexandercerutti/passkit-generator/wiki/Troubleshooting-(Self-help)#through-console-mac-only

Let me know.

**Reply by mbappai** [reply]
Alright let me do that and see how it works

**Reply by mbappai** [reply]
okay so I just used the console app and I'm getting the following error the moment I click on "Add to Apple pass" button on my mobile safarI
`Invalid data error reading pass pass.com.flexabledats.marketplace/nmyuxofgna. Manifest signature did not verify successfully`

**Reply by alexandercerutti** [reply]
Uh, there must be something wrong with your certicates. That's actually a real mess to understand what exactly went wrong when this happens...

**Reply by mbappai** [reply]
Huh! Alright let me go over the step again

**Reply by mbappai** [reply]
Ahh! I found the issue. So the problem was at step 6 of generating certificates through terminal where I was prompted for information to generate the CSR. Instead of me to provide Apple CA's information, I now went ahead and asssumed they were placeholders and just filled my company's information. 

I'm finally able to get it to work after making that change. 

Thanks a lot for your help Alex.

After I'm through with the integration in my project I'd really like to contribute to the documentation a little bit if that's okay

**Reply by alexandercerutti** [reply]
@mbappai So you inserted your information instead of Apple's or skipped them at all?

Just to understand, because I want to be honest with you: I started thinking those informations to fill when creating the CSR were quite useless because sometimes I generated the certificates again without them (going straight as I understood you did) and it seemed to work.

So, I actually don't know.

Anyway, I'm really glad you were able to solve!

Yeah sure, feel free to open a PR or an issue and we'll discuss about the changes ;)

**Reply by mbappai** [reply]
hahaha Sorry my message wasn't clear; I inserted Apple's information as I previously had my information which didn't work for me. 

So may be, it's either you provide Apple's information or no information at all (According to what you did). That part needs to be tested a little bit more to confirm an intended behaviour I guess.

Yeah will certainly be opening those PR's. Thanks again.

Cheers!
